File: sdk/python/marketarena.py
import urllib.request
import urllib.parse
import json
import hmac
import hashlib
import time

class MarketArenaClient:
    """
    MarketArena Python Client SDK
    Pure standard library client for algorithmic trading and quantitative research.
    """
    def __init__(self, base_url="http://localhost:3000", api_key=None, api_secret=None):
        self.base_url = base_url.rstrip('/')
        self.api_key = api_key
        self.api_secret = api_secret

    def _request(self, method, endpoint, body=None):
        url = f"{self.base_url}{endpoint}"
        headers = {
            "Accept": "application/json",
            "User-Agent": "MarketArena-Python-SDK/0.8"
        }

        method_upper = method.upper()
        data_bytes = None

        if body is not None and method_upper != "GET":
            body_str = json.dumps(body)
            data_bytes = body_str.encode('utf-8')
            headers["Content-Type"] = "application/json"
        else:
            body_str = ""

        if self.api_key and self.api_secret:
            timestamp = str(int(time.time() * 1000))
            payload = f"{timestamp}{method_upper}{endpoint}{body_str}"
            signature = hmac.new(
                self.api_secret.encode('utf-8'),
                payload.encode('utf-8'),
                hashlib.sha256
            ).hexdigest()

            headers["X-API-KEY"] = self.api_key
            headers["X-API-TIMESTAMP"] = timestamp
            headers["X-API-SIGNATURE"] = signature
        elif self.api_key:
            headers["X-API-KEY"] = self.api_key

        req = urllib.request.Request(url, data=data_bytes, headers=headers, method=method_upper)
        try:
            with urllib.request.urlopen(req) as resp:
                resp_data = resp.read().decode('utf-8')
                return json.loads(resp_data)
        except urllib.error.HTTPError as err:
            err_data = err.read().decode('utf-8')
            try:
                parsed_err = json.loads(err_data)
                msg = parsed_err.get("message") or parsed_err.get("error") or str(err)
            except Exception:
                msg = str(err)
            raise RuntimeError(f"HTTP {err.code}: {msg}")

    # --- Public Market Data ---

    def ping(self):
        """Server health check and market clock state"""
        return self._request("GET", "/api/v1/ping")

    def get_regime(self):
        """Get active macroeconomic market regime"""
        return self._request("GET", "/api/v1/regime")

    def get_orderbook(self, symbol, depth=10):
        """Fetch L2 depth bids and asks"""
        return self._request("GET", f"/api/v1/orderbook/{symbol}?depth={depth}")

    def get_candles(self, symbol, timeframe="5s", limit=50):
        """
        Fetch historical OHLCV candlestick bars
        Timeframes: '1s', '5s', '15s', '1m', '5m'
        """
        return self._request("GET", f"/api/v1/candles/{symbol}?timeframe={timeframe}&limit={limit}")

    # --- Authenticated Endpoints ---

    def get_account(self):
        """Fetch account balances, portfolio valuation, and risk metrics"""
        return self._request("GET", "/api/v1/account")

    def get_open_orders(self):
        """List active resting orders"""
        return self._request("GET", "/api/v1/orders")

    def place_order(self, symbol, side, quantity, order_type="LIMIT", price=None, stop_price=None, trailing_delta=None, leverage=1, is_short=False):
        """
        Submit order to matching engine
        """
        payload = {
            "symbol": symbol,
            "side": side.upper(),
            "type": order_type.upper(),
            "quantity": quantity,
            "leverage": leverage,
            "isShort": bool(is_short)
        }
        if price is not None:
            payload["price"] = price
        if stop_price is not None:
            payload["stopPrice"] = stop_price
        if trailing_delta is not None:
            payload["trailingDelta"] = trailing_delta

        return self._request("POST", "/api/v1/orders", payload)

    def cancel_order(self, symbol, order_id):
        """Cancel resting order"""
        return self._request("DELETE", f"/api/v1/orders/{order_id}?symbol={symbol}")

    @staticmethod
    def to_dataframe(candle_response):
        """Convert candle data to pandas DataFrame if pandas is installed"""
        try:
            import pandas as pd
            candles = candle_response.get("candles", [])
            df = pd.DataFrame(candles)
            if not df.empty and "time" in df.columns:
                df["time"] = pd.to_datetime(df["time"], unit="ms")
            return df
        except ImportError:
            raise ImportError("pandas is required for to_dataframe(). Run: pip install pandas")

File: sdk/js/marketarena.js
import crypto from 'crypto';

/**
 * MarketArena JavaScript/Node.js Client SDK
 * Programmatic trading and market data interface for external bots.
 */
export class MarketArenaClient {
  /**
   * @param {Object} options
   * @param {string} [options.baseUrl='http://localhost:3000']
   * @param {string} [options.apiKey]
   * @param {string} [options.apiSecret]
   */
  constructor({ baseUrl = 'http://localhost:3000', apiKey = null, apiSecret = null } = {}) {
    this.baseUrl = baseUrl.replace(/\/+$/, '');
    this.apiKey = apiKey;
    this.apiSecret = apiSecret;
  }

  /**
   * Internal authenticated request dispatcher
   */
  async _request(method, endpoint, body = null) {
    const url = `${this.baseUrl}${endpoint}`;
    const headers = {
      'Accept': 'application/json'
    };

    const methodUpper = method.toUpperCase();
    let bodyStr = '';

    if (body !== null && methodUpper !== 'GET') {
      bodyStr = JSON.stringify(body);
      headers['Content-Type'] = 'application/json';
    }

    if (this.apiKey && this.apiSecret) {
      const timestamp = Date.now().toString();
      const payload = `${timestamp}${methodUpper}${endpoint}${bodyStr}`;
      const signature = crypto.createHmac('sha256', this.apiSecret).update(payload).digest('hex');

      headers['X-API-KEY'] = this.apiKey;
      headers['X-API-TIMESTAMP'] = timestamp;
      headers['X-API-SIGNATURE'] = signature;
    } else if (this.apiKey) {
      headers['X-API-KEY'] = this.apiKey;
    }

    const res = await fetch(url, {
      method: methodUpper,
      headers,
      body: bodyStr || undefined
    });

    const data = await res.json().catch(() => ({}));
    if (!res.ok) {
      const err = new Error(data.message || data.error || `HTTP ${res.status}: ${res.statusText}`);
      err.status = res.status;
      err.response = data;
      throw err;
    }

    return data;
  }

  // --- Public Market Data ---

  /**
   * Server ping & health check
   */
  async ping() {
    return this._request('GET', '/api/v1/ping');
  }

  /**
   * Get active macroeconomic market regime
   */
  async getRegime() {
    return this._request('GET', '/api/v1/regime');
  }

  /**
   * Get L2 order book depth for a symbol
   * @param {string} symbol (e.g. 'BYTE')
   * @param {number} [depth=10]
   */
  async getOrderbook(symbol, depth = 10) {
    return this._request('GET', `/api/v1/orderbook/${symbol}?depth=${depth}`);
  }

  /**
   * Get historical OHLCV candlestick bars
   * @param {string} symbol (e.g. 'BYTE')
   * @param {string} [timeframe='5s'] '1s'|'5s'|'15s'|'1m'|'5m'
   * @param {number} [limit=50]
   */
  async getCandles(symbol, timeframe = '5s', limit = 50) {
    return this._request('GET', `/api/v1/candles/${symbol}?timeframe=${timeframe}&limit=${limit}`);
  }

  // --- Authenticated Account & Trading Endpoints ---

  /**
   * Get authenticated user account details, portfolio valuation, and risk metrics
   */
  async getAccount() {
    return this._request('GET', '/api/v1/account');
  }

  /**
   * Get authenticated user active resting orders
   */
  async getOpenOrders() {
    return this._request('GET', '/api/v1/orders');
  }

  /**
   * Submit an order to the matching engine
   * @param {Object} order
   * @param {string} order.symbol
   * @param {'BUY'|'SELL'} order.side
   * @param {'LIMIT'|'MARKET'|'STOP_LOSS'|'STOP_LIMIT'|'TRAILING_STOP'} [order.type='LIMIT']
   * @param {number} order.quantity
   * @param {number} [order.price] Required for LIMIT/STOP_LIMIT
   * @param {number} [order.stopPrice] Required for STOP_LOSS/STOP_LIMIT
   * @param {number} [order.trailingDelta] Required for TRAILING_STOP
   * @param {number} [order.leverage=1] Margin leverage (1, 2, or 5)
   * @param {boolean} [order.isShort=false] Short sell execution
   */
  async placeOrder(order) {
    return this._request('POST', '/api/v1/orders', order);
  }

  /**
   * Cancel a resting order
   * @param {string} symbol
   * @param {string} orderId
   */
  async cancelOrder(symbol, orderId) {
    return this._request('DELETE', `/api/v1/orders/${orderId}?symbol=${symbol}`);
  }

  /**
   * Generate an API key pair for a user
   * @param {string} userId
   * @param {string} [name='Bot Key']
   * @param {Array<string>} [permissions=['read', 'trade']]
   */
  async createApiKey(userId, name = 'Bot Key', permissions = ['read', 'trade']) {
    return this._request('POST', '/api/v1/keys', { userId, name, permissions });
  }

  /**
   * List API keys for a user
   * @param {string} userId
   */
  async getUserApiKeys(userId) {
    return this._request('GET', `/api/v1/keys?userId=${userId}`);
  }
}
